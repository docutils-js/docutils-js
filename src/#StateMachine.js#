class StateMachine {
    /* 
        Initialize a `StateMachine` object; add state objects.

       Parameters:

        - `state_classes`: a list of `State` (sub)classes.
        - `initial_state`: a string, the class name of the initial state.
        - `debug`: a boolean; produce verbose output if true (nonzero).
	*/

    constructor(stateClasses, initialState, debug) {
	if(debug === undefined) {
	    debug = false;
	}
	this.inputLines = undefined;
	this.inputOffset = 0;
	this.line = undefined;
	this.lineOffset = -1;
	this.debug = debug;
	this.initialState = initialState;
	this.currentState = initialState;
	this.states = {};
	this.addStates(stateClasses);
	this.observers = []
	//       self._stderr = ErrorOutput()
    }

    unlink() {
	Object.values(this.states).forEach(s => s.unlink());
	this.states = undefined;
    }

    run(inputLines, inputOffset = 0, context, inputSource, initialState) {
	/*
        Run the state machine on `input_lines`. Return results (a list).

        Reset `self.line_offset` and `self.current_state`. Run the
        beginning-of-file transition. Input one line at a time and check for a
        matching transition. If a match is found, call the transition method
        and possibly change the state. Store the context returned by the
        transition method to be passed on to the next transition matched.
        Accumulate the results returned by the transition methods in a list.
        Run the end-of-file transition. Finally, return the accumulated
        results.

        Parameters:

        - `input_lines`: a list of strings without newlines, or `StringList`.
        - `input_offset`: the line offset of `input_lines` from the beginning
          of the file.
        - `context`: application-specific storage.
        - `input_source`: name or path of source of `input_lines`.
        - `initial_state`: name of initial state.
	*/
	this.runtimeInit();
	//if isinstance(input_lines, StringList):
	//self.input_lines = input_lines
	//else:
	//self.input_lines = StringList(input_lines, source=input_source)
	this.inputLines = inputLines;
	this.lineOffset = -1;
	this.currentState = initialState || this.initialState;
	let transitions = undefined;
	const results = [];
	let state = this.getState();
	let nextState;
	try {
	    [ context, result ] = state.bof(context);
	    results.push(result);
	    while(true) {
		try {
		    try {
			this.nextLine();
			[ context, nextState, result ] = this.checkLine(context, state, transitions);
		    }
		    catch(error) {
		    }
		}
		catch(error) {
		}
	    }
	}
	catch(error) {
	}
	this.observers = [];
	return results;
    }

    addStates(stateClasses) {
	// stub
    }

    runtimeInit() {
	Object.values(this.states).forEach(s => s.runtimeInit());
    }

    getState(nextState) {
	if(nextState) {
	    this.currentState = nextState;
	}
	
	    
}

export default StateMachine;
