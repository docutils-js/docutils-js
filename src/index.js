export const __version__ = '0.14js';
