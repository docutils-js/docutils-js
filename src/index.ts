/* eslint-disable-next-line import/prefer-default-export */
export { parse } from './parse';
export { StringOutput, StringInput } from './io';
export { StandaloneReader } from './readers/standalone';
export { newDocument } from './newDocument';
import * as nodes from './nodes';
import Writer from './Writer';
import Transform from './Transform';
export { Reader } from './Reader';
export { Publisher } from './Publisher'
export { defaults } from '../gen/defaults';
export { pojoTranslate } from './fn/pojoTranslate';
export { htmlTranslate } from './fn/htmlTranslate';
export { RSTParser } from './parsers/restructuredtext';
export { XMLWriter } from './writers/xml'

/* eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars */
const __docformat__ = 'reStructuredText';

export const __version__ = '0.14js';

export {
    nodes, Writer, Transform,
};
