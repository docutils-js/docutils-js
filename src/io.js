import TransformSpec  from './TransformSpec';
import { createReadStream } from 'fs';
import ErrorOutput from './ErrorOutput';
import Input from './io/Input';
import Output from './io/Output';
import { ApplicationError } from'./Exceptions';

export class StringInput extends Input {
    constructor(...args) {
	super(...args);
	this.sourcePath = '<string>';
    }
	
    read(cb) {
	cb(undefined, this.source);
    }
}

export class StringOutput extends Output {
    constructor(...args) {
	super(...args);
	this.defaultDestinationPath = '<string>'
    }
    
    write(data) {
        //self.destination = self.encode(data) // fixme encoding
	if(Array.isArray(data)) {
	    data = JSON.stringify(data);
	}
	this.destination = data;
	return this.destination;
    }
}

export class FileInput extends Input {
    /* ew, too much logic for a constructor, with side effects etc! */
    constructor(args) {
	super(args);
	let { source, sourcePath, encoding, errorHandler, autoClose,
	      mode } = args;
	if(typeof source === 'undefined' && typeof sourcePath === 'undefined') {
	    throw new ApplicationError("FileInput: Undefined source and sourcePath");
	}
	
	if(autoClose === undefined) {
	    autoClose = true;
	}
	if(mode === undefined) {
	    mode = 'r';
	}
	
	this.autoClose = autoClose;
	this._stderr = new ErrorOutput()
	if(!source) {
	    if(sourcePath) {
		try {
		    this.source = createReadStream(sourcePath, { encoding: 'utf8' });
		}
		catch(error) {
		    console.log(error.stack);
		    throw error;
		}
	    } else {
		this.source = process.stdin;
	    }
	} else {
	    
	    // ??
	}
	if(!sourcePath) {
	    this.sourcePath = this.source.name;
	}
    }

    /* Read and decode a single file and return the data (Unicode string).
     */
    read(cb) {
	setTimeout(() => {
	let data;
	try {
	    /* reading ? */
	    if(this.source === process.stdin) {
		// do stuff
	    } else {
		data = this.source.read();
		if(data === null) {
		    console.log("read returned null ?");
		}
		cb(undefined, data);
	    }
	} catch(error) {
	    console.log(error.stack);
	    cb(error, undefined);
	}
	}, 100);
    }

    readLines() {
	return this.read().splitlines(true);
    }

    close() {
	if(this.source !== process.stdin) {
	    this.source.close();
	}
    }
}

export class FileOutput extends Output {
    
}
