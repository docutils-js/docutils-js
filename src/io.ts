import Input from './io/Input';
import Output from './io/Output';
import { ReadInputCallback, LoggerType } from "./types";

/** Direct string input. */
export class StringInput extends Input {
    public constructor(source: string, logger: LoggerType, sourcePath?: string, encoding?: string, errorHandler?: string) {
        super({source, sourcePath, encoding, errorHandler, logger});
        this.sourcePath = '<string>';
    }

    public read(cb: ReadInputCallback<string|string[]|{}>): void {
        cb(undefined, this.source);
    }
}

export class StringOutput extends Output<string> {
    public constructor(
        logger: LoggerType,
        destination?: string,
        destinationPath?: string,
        encoding?: string,
        errorHandler?: string
    ) {
        super(logger, destination, destinationPath, encoding, errorHandler);
        this.defaultDestinationPath = '<string>';

    }

    public write(data: string): string {
        // self.destination = self.encode(data) // fixme encoding
        if (Array.isArray(data)) {
            data = JSON.stringify(data);
        }
        this.destination = data;
        return this.destination;
    }
}

