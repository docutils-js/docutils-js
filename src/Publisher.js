import { FileInput, FileOutput } from './io';
import { ApplicationError } from'./Exceptions';
import readers from './Readers'
import writers from './Writers'

class Publisher {
    constructor(args) {
	let { reader, parser, writer, source, sourceClass, destination,
	      destinationClass, settings, debugFn }  = args;
	console.log(source);
	/* Terrible defaults ! */
	if(!sourceClass) {
	    sourceClass = FileInput;
	}
	if(!destinationClass) {
	    destinationClass = FileOutput;
	}
	this.debugFn = debugFn;
	this.document = null;
	this.reader = reader;
	this.parser = parser;
	this.writer = writer;
	this.source = source;
	if(sourceClass === undefined) {
	    this.sourceClass = Source;
	} else {
	    this.sourceClass = sourceClass;
	}

	this.destination = destination;
	if(destinationClass === undefined) {
	    this.destinationClass = Destination;
	}
	else {
	    this.destinationClass = destinationClass;
	}
	this.settings = settings;
    }

    setReader(readerName, parser, parserName) {
	const ReaderClass = readers.getReaderClass(readerName)
	this.reader = new ReaderClass(parser, parserName, { debug: this.debug, debugFn: this.debugFn })
	this.parser = this.reader.parser
    }

    setWriter(writerName) {
	const writerClass = writers.getWriterClass(writerName)
	this.writer = new writerClass();
    }

    setComponents(readerName, parserName, writerName) {
	if(!this.reader) {
	    this.setReader(readerName, this.parser, parserName);
	}
	if(!this.parser) {
	    if(!this.reader.parsser) {
                this.reader.setParser(parserName)
	    }
	    this.parser = this.reader.parser
	}
	if(!this.writer) {
            this.setWriter(writerName)
	}
    }

    setupOptionParser({usage, description, settingsSpec, configSection, defaults}) {
	if(configSection) {
	    if(!settingsSpec) {
		settingsSpec = new SettingsSpec();
	    }
	    settingsSpec.configSection = configSection;
	    const parts = configSection.split();
	    if(parts.length > 1 && parts[parts.length - 1] === 'application') {
		settingsSpec.configSectionDepenendencies = ['applications']

	    }
	}
	const optionParser = new OptionParser({components: [this.parser, this.reader, this.writer, settingsSpec],
					       defaults,
					       readConfigFiles: true,
					       usage,description });
//	console.log(JSON.stringify(optionParser.settingsSpec));
	return optionParser;
    }
    
    processCommandLine({argv, usage, description, settingsSpec, configSection,
			settingsOverrides}) {
	const optionParser = this.setupOptionParser({usage,description,settingsSpec,
					       configSection, settingsOverrides});
	if(argv === undefined) {
	    argv = process.argv.slice(2);
	}
	this.settings = optionParser.parseArgs(argv);
	// bad! fixme
	this.settings.idPrefix = '';
	this.settings.autoIdPrefix = 'auto'
	this.settings.debug = true; // force debug true for now
    }

    setIO(sourcePath, destinationPath) {
	if(typeof this.source === 'undefined') {
	    this.setSource({sourcePath});
	}
	if(this.destination === undefined) {
	    this.setDestination({destinationPath});
	}
    }

    setSource({source, sourcePath}) {
	if(sourcePath === undefined) {
	    sourcePath = this.settings._source;
	}else {
	    this.settings._source = sourcePath;
	}
	try {
	    const sourceClass = this.sourceClass;
	    this.source = new sourceClass({source, sourcePath, encoding:
						this.settings.inputEncoding});
	} catch(error) {
	    throw new ApplicationError(`Unable to instantiate Source class ${this.sourceClass.constructor.name}: ${error.message}`);
	}
    }

    setDestination({ destination, destinationPath}) {
	if(destinationPath === undefined) {
	    destinationPath = this.settings._destination;
	}else{
	    this.settings._destination= destinationPath;
	}
	const destinationClass = this.destinationClass;
	this.destination = new destinationClass({destination,
						  destinationPath,
						  encoding: this.settings.outputEncoding,
						  errorHandler: this.settings.outputEncodingErrorHandler })
    }

    applyTransforms() {
	this.document.transformer.populateFromComponents(this.source, this.reader, this.reader.parser, this.writer, this.destination);
	this.document.transformer.applyTransforms();
    }

    /* This doesnt seem to return anything ? */
    publish(args, cb) {
//	console.log(`publish`)
	const {argv, usage, description, settingsSpec, settingsOverrides, configSection, enableExitStatus } = args;
	let exit = undefined;
	try {
	    if(this.settings === undefined) {
		this.processCommandLine({ argv, usage, description, settingsSpec, configSection, settingsOverrides});
	    }
	    console.log(this.source);
	    this.setIO();

	    //KM
//	    console.log('*** about to call read');
	    /* we may need to change semantics here !! */

	    if(typeof this.reader === 'undefined') {
		throw new ApplicationError('Need defined reader with "read" method');
	    }
	    this.reader.read(
		this.source, this.parser, this.settings,
		((error, document) => {
		    if(error) {
			cb(error, undefined);
			return;
		    }
		    this.document = document;
		    if(!document) {
			throw new Error("need document");
		    }
		    this.applyTransforms()
		    const output =
			  this.writer.write(this.document, this.destination)
		    this.writer.assembleParts();
		    cb(undefined, output);
		}).bind(this));
	} catch(error) {
	    cb(error, undefined);
	}
    }
}

export default Publisher;
