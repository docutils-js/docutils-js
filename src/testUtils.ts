import { newDocument } from './newDocument';
import newReporter    from './newReporter';
import { defaultPublisherOptions } from '../src/constants';
import { getDefaultSettings } from './settingsHelper';
import winston, { format, transports, LoggerOptions, Logger } from 'winston';
import { defaultConsoleLogLevel } from './constants';
import {Publisher} from './Publisher';
import path from 'path';
import StateFactory from './parsers/rst/StateFactory';
import { LoggerType } from './types';
import process from 'process';

let logger: LoggerType|undefined;

export function createLogger(options?: LoggerOptions): Logger {
    if(logger !== undefined) {
        return logger;
    }
    let myOpt = options === undefined ? {} : { ...options};
    if(myOpt.format === undefined) {
        myOpt.format = format.json();
    }
    if(myOpt.transports === undefined) {
        myOpt.transports = [
            new transports.Console({level: defaultConsoleLogLevel}),
            new transports.File({filename: `${path.basename(process.argv[1])}-${process.pid}.log`, level: 'silly'}),
        ];
    }
    logger = winston.createLogger(myOpt);
    return logger;
}

export function createPublisher() {
    const publisher = new Publisher({ ...defaultPublisherOptions, logger: createLogger() });
    publisher.setComponents(defaultPublisherOptions.readerName, defaultPublisherOptions.parserName, 'xml');
    return publisher;
}

export function createStateMachine() {
}

export function createStateFactory(): StateFactory {
return new StateFactory({ logger: createLogger() });
}

export function createNewDocument(sourcePath: string='default') {
    return newDocument({sourcePath, logger: createLogger()}, getDefaultSettings());
}
  
export function createNewReporter(sourcePath: string='default') {
    return newReporter({sourcePath}, getDefaultSettings());
}
