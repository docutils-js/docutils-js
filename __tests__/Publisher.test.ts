import {Publisher} from '../src/Publisher';
import { StringInput } from'../src/io';
import { createPublisher,createLogger } from '../src/testUtils';
import sinon from 'sinon';

test.skip('1',() => {
const logger = createLogger();
const input = new StringInput('', logger);
sinon.spy(input, 'read');
const p = createPublisher();
// @ts-ignore
expect(input.read.calledOnce).toBeTruthy();
});
