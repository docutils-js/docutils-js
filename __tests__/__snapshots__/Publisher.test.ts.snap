// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Instantiate publisher #2 1`] = `
document {
  "attributes": Object {
    "backrefs": Array [],
    "classes": Array [],
    "dupnames": Array [],
    "ids": Array [],
    "names": Array [],
    "source": "",
  },
  "autofootnoteRefs": Array [],
  "autofootnoteStart": 1,
  "autofootnotes": Array [],
  "basicAttributes": Array [
    "ids",
    "classes",
    "names",
    "dupnames",
  ],
  "childTextSeparator": "

",
  "children": Array [],
  "citationRefs": Object {},
  "citations": Array [],
  "classTypes": Array [
    [Function],
    [Function],
  ],
  "currentLine": 0,
  "currentSource": "",
  "decoration": undefined,
  "document": [Circular],
  "footnoteRefs": Object {},
  "footnotes": Array [],
  "idStart": 1,
  "ids": Object {},
  "indirectTargets": Array [],
  "knownAttributes": Array [
    "ids",
    "classes",
    "names",
    "dupnames",
    "backrefs",
    "source",
    "rawsource",
  ],
  "line": 0,
  "listAttributes": Array [
    "ids",
    "classes",
    "names",
    "dupnames",
    "backrefs",
  ],
  "localAttributes": Array [
    "backrefs",
  ],
  "nameIds": Object {},
  "nameTypes": Object {},
  "names": Array [],
  "nodeName": Symbol(Element),
  "parseMessages": Array [],
  "rawsource": "",
  "refIds": Object {},
  "refNames": Object {},
  "referenced": false,
  "reporter": Reporter {
    "DEBUG_LEVEL": 0,
    "ERROR_LEVEL": 3,
    "INFO_LEVEL": 1,
    "SEVERE_LEVEL": 4,
    "WARNING_LEVEL": 2,
    "debugFlag": undefined,
    "debugLevel": 0,
    "encoding": "UTF-8",
    "errorHandler": "backslashreplace",
    "haltLevel": 4,
    "maxLevel": -1,
    "observers": Array [
      [Function],
    ],
    "reportLevel": undefined,
    "source": "",
    "stream": undefined,
  },
  "settings": Object {
    "docutilsCoreOptionParser": Object {
      "autoIdPrefix": "id",
      "debug": undefined,
      "errorEncoding": "UTF-8",
      "errorEncodingErrorHandler": "backslashreplace",
      "exitStatusLevel": 5,
      "haltLevel": undefined,
      "idPrefix": "",
      "inputEncodingErrorHandler": "strict",
      "languageCode": "en",
      "outputEncoding": "utf-8",
      "outputEncodingErrorHandler": "strict",
      "recordDependencies": null,
      "reportLevel": undefined,
    },
    "docutilsFrontendOptionParser": Object {
      "autoIdPrefix": "id",
      "errorEncoding": "UTF-8",
      "errorEncodingErrorHandler": "backslashreplace",
      "exitStatusLevel": 5,
      "idPrefix": "",
      "inputEncodingErrorHandler": "strict",
      "languageCode": "en",
      "outputEncoding": "utf-8",
      "outputEncodingErrorHandler": "strict",
      "recordDependencies": null,
    },
    "docutilsParsersRstParser": Object {
      "characterLevelInlineMarkup": false,
      "pepBaseUrl": "http://www.python.org/dev/peps/",
      "pepFileUrlTemplate": "pep-%04d",
      "rfcBaseUrl": "http://tools.ietf.org/html/",
      "smartQuotes": false,
      "syntaxHighlight": "long",
      "tabWidth": 8,
    },
    "docutilsReadersPepReader": Object {},
    "docutilsReadersStandaloneReader": Object {
      "docinfoXform": 1,
      "doctitleXform": 1,
    },
    "docutilsWritersDocutilsXmlWriter": Object {
      "doctypeDeclaration": 1,
      "xmlDeclaration": 1,
    },
    "docutilsWritersHtml4Css1Writer": Object {
      "attribution": "dash",
      "fieldNameLimit": 14,
      "footnoteReferences": "brackets",
      "initialHeaderLevel": "1",
      "mathOutput": "HTML math.css",
      "optionLimit": 14,
      "stylesheetDirs": Array [
        ".",
        "/usr/share/docutils/writers/html4css1",
        "/usr/share/docutils/writers/html5_polyglot",
      ],
      "stylesheetPath": Array [
        "html4css1.css",
      ],
      "tableStyle": "",
      "template": "/usr/share/docutils/writers/html4css1/template.txt",
      "xmlDeclaration": 1,
    },
    "docutilsWritersHtml5PolyglotWriter": Object {
      "attribution": "dash",
      "footnoteReferences": "brackets",
      "initialHeaderLevel": "1",
      "mathOutput": "HTML math.css",
      "stylesheetDirs": Array [
        ".",
        "/usr/share/docutils/writers/html5_polyglot",
      ],
      "stylesheetPath": Array [
        "minimal.css",
        "plain.css",
      ],
      "tableStyle": "",
      "template": "/usr/share/docutils/writers/html5_polyglot/template.txt",
    },
    "docutilsWritersLatex2EWriter": Object {
      "attribution": "dash",
      "documentclass": "article",
      "documentoptions": "a4paper",
      "docutilsFootnotes": true,
      "embedStylesheet": 0,
      "fontEncoding": "T1",
      "footnoteReferences": "superscript",
      "graphicxOption": "",
      "hyperlinkColor": "blue",
      "hyperrefOptions": "",
      "latexPreamble": "% PDF Standard Fonts
\\\\usepackage{mathptmx} % Times
\\\\usepackage[scaled=.90]{helvet}
\\\\usepackage{courier}",
      "literalBlockEnv": "",
      "referenceLabel": null,
      "sectionEnumeratorSeparator": "-",
      "stylesheet": "",
      "stylesheetDirs": Array [
        ".",
      ],
      "tableStyle": Array [
        "standard",
      ],
      "template": "default.tex",
      "useBibtex": null,
      "useLatexAbstract": 0,
      "useLatexDocinfo": 0,
      "usePartSection": 0,
      "useVerbatimWhenPossible": 0,
    },
    "docutilsWritersOdfOdtReader": Object {
      "docinfoXform": 1,
      "doctitleXform": 1,
    },
    "docutilsWritersOdfOdtWriter": Object {
      "addSyntaxHighlighting": false,
      "cloakEmailAddresses": false,
      "createLinks": false,
      "createSections": true,
      "customFooter": "",
      "customHeader": "",
      "endnotesEndDoc": false,
      "generateOowriterToc": true,
      "stylesheet": "/usr/share/docutils/writers/odf_odt/styles.odt",
      "tableBorderThickness": null,
    },
    "docutilsWritersPepHtmlWriter": Object {
      "attribution": "dash",
      "fieldNameLimit": 14,
      "footnoteReferences": "brackets",
      "initialHeaderLevel": "1",
      "mathOutput": "HTML math.css",
      "optionLimit": 14,
      "stylesheetDirs": Array [
        ".",
        "/usr/share/docutils/writers/html4css1",
        "/usr/share/docutils/writers/html5_polyglot",
      ],
      "stylesheetPath": Array [
        "html4css1.css",
      ],
      "tableStyle": "",
      "template": "/usr/share/docutils/writers/html4css1/template.txt",
      "xmlDeclaration": 1,
    },
    "docutilsWritersS5HtmlWriter": Object {
      "attribution": "dash",
      "fieldNameLimit": 14,
      "footnoteReferences": "brackets",
      "initialHeaderLevel": "1",
      "mathOutput": "HTML math.css",
      "optionLimit": 14,
      "stylesheetDirs": Array [
        ".",
        "/usr/share/docutils/writers/html4css1",
        "/usr/share/docutils/writers/html5_polyglot",
      ],
      "stylesheetPath": Array [
        "html4css1.css",
      ],
      "tableStyle": "",
      "template": "/usr/share/docutils/writers/html4css1/template.txt",
      "xmlDeclaration": 1,
    },
    "docutilsWritersXetexWriter": Object {
      "attribution": "dash",
      "documentclass": "article",
      "documentoptions": "a4paper",
      "docutilsFootnotes": true,
      "embedStylesheet": 0,
      "footnoteReferences": "superscript",
      "graphicxOption": "",
      "hyperlinkColor": "blue",
      "hyperrefOptions": "",
      "latexPreamble": "% Linux Libertine (free, wide coverage, not only for Linux)
\\\\setmainfont{Linux Libertine O}
\\\\setsansfont{Linux Biolinum O}
\\\\setmonofont[HyphenChar=None,Scale=MatchLowercase]{DejaVu Sans Mono}",
      "literalBlockEnv": "",
      "referenceLabel": null,
      "sectionEnumeratorSeparator": "-",
      "stylesheet": "",
      "stylesheetDirs": Array [
        ".",
      ],
      "tableStyle": Array [
        "standard",
      ],
      "template": "xelatex.tex",
      "useBibtex": null,
      "useLatexAbstract": 0,
      "useLatexDocinfo": 0,
      "usePartSection": 0,
      "useVerbatimWhenPossible": 0,
    },
  },
  "source": "",
  "substitutionDefs": Object {},
  "substitutionNames": Object {},
  "symbolFootnoteRefs": Array [],
  "symbolFootnoteStart": 0,
  "symbolFootnotes": Array [],
  "tagname": "document",
  "transformMessages": Array [],
  "transformer": Transformer {
    "applied": Array [],
    "components": Object {
      "undefined": Output {
        "_init": [MockFunction],
        "encode": [MockFunction],
        "getTransforms": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "toString": [MockFunction],
        "write": [MockFunction],
      },
    },
    "document": [Circular],
    "serialno": 0,
    "sorted": 0,
    "transforms": Array [],
    "unknownReferenceResolvers": Array [],
  },
}
`;

exports[`Instantiate publisher 1`] = `
document {
  "attributes": Object {
    "backrefs": Array [],
    "classes": Array [],
    "dupnames": Array [],
    "ids": Array [],
    "names": Array [],
    "source": "",
  },
  "autofootnoteRefs": Array [],
  "autofootnoteStart": 1,
  "autofootnotes": Array [],
  "basicAttributes": Array [
    "ids",
    "classes",
    "names",
    "dupnames",
  ],
  "childTextSeparator": "

",
  "children": Array [],
  "citationRefs": Object {},
  "citations": Array [],
  "classTypes": Array [
    [Function],
    [Function],
  ],
  "currentLine": 0,
  "currentSource": "",
  "decoration": undefined,
  "document": [Circular],
  "footnoteRefs": Object {},
  "footnotes": Array [],
  "idStart": 1,
  "ids": Object {},
  "indirectTargets": Array [],
  "knownAttributes": Array [
    "ids",
    "classes",
    "names",
    "dupnames",
    "backrefs",
    "source",
    "rawsource",
  ],
  "line": 0,
  "listAttributes": Array [
    "ids",
    "classes",
    "names",
    "dupnames",
    "backrefs",
  ],
  "localAttributes": Array [
    "backrefs",
  ],
  "nameIds": Object {},
  "nameTypes": Object {},
  "names": Array [],
  "nodeName": Symbol(Element),
  "parseMessages": Array [],
  "rawsource": "",
  "refIds": Object {},
  "refNames": Object {},
  "referenced": false,
  "reporter": Reporter {
    "DEBUG_LEVEL": 0,
    "ERROR_LEVEL": 3,
    "INFO_LEVEL": 1,
    "SEVERE_LEVEL": 4,
    "WARNING_LEVEL": 2,
    "debugFlag": undefined,
    "debugLevel": 0,
    "encoding": "UTF-8",
    "errorHandler": "backslashreplace",
    "haltLevel": 4,
    "maxLevel": -1,
    "observers": Array [
      [Function],
    ],
    "reportLevel": undefined,
    "source": "",
    "stream": undefined,
  },
  "settings": Object {
    "docutilsCoreOptionParser": Object {
      "autoIdPrefix": "id",
      "debug": undefined,
      "errorEncoding": "UTF-8",
      "errorEncodingErrorHandler": "backslashreplace",
      "exitStatusLevel": 5,
      "haltLevel": undefined,
      "idPrefix": "",
      "inputEncodingErrorHandler": "strict",
      "languageCode": "en",
      "outputEncoding": "utf-8",
      "outputEncodingErrorHandler": "strict",
      "recordDependencies": null,
      "reportLevel": undefined,
    },
    "docutilsFrontendOptionParser": Object {
      "autoIdPrefix": "id",
      "errorEncoding": "UTF-8",
      "errorEncodingErrorHandler": "backslashreplace",
      "exitStatusLevel": 5,
      "idPrefix": "",
      "inputEncodingErrorHandler": "strict",
      "languageCode": "en",
      "outputEncoding": "utf-8",
      "outputEncodingErrorHandler": "strict",
      "recordDependencies": null,
    },
    "docutilsParsersRstParser": Object {
      "characterLevelInlineMarkup": false,
      "pepBaseUrl": "http://www.python.org/dev/peps/",
      "pepFileUrlTemplate": "pep-%04d",
      "rfcBaseUrl": "http://tools.ietf.org/html/",
      "smartQuotes": false,
      "syntaxHighlight": "long",
      "tabWidth": 8,
    },
    "docutilsReadersPepReader": Object {},
    "docutilsReadersStandaloneReader": Object {
      "docinfoXform": 1,
      "doctitleXform": 1,
    },
    "docutilsWritersDocutilsXmlWriter": Object {
      "doctypeDeclaration": 1,
      "xmlDeclaration": 1,
    },
    "docutilsWritersHtml4Css1Writer": Object {
      "attribution": "dash",
      "fieldNameLimit": 14,
      "footnoteReferences": "brackets",
      "initialHeaderLevel": "1",
      "mathOutput": "HTML math.css",
      "optionLimit": 14,
      "stylesheetDirs": Array [
        ".",
        "/usr/share/docutils/writers/html4css1",
        "/usr/share/docutils/writers/html5_polyglot",
      ],
      "stylesheetPath": Array [
        "html4css1.css",
      ],
      "tableStyle": "",
      "template": "/usr/share/docutils/writers/html4css1/template.txt",
      "xmlDeclaration": 1,
    },
    "docutilsWritersHtml5PolyglotWriter": Object {
      "attribution": "dash",
      "footnoteReferences": "brackets",
      "initialHeaderLevel": "1",
      "mathOutput": "HTML math.css",
      "stylesheetDirs": Array [
        ".",
        "/usr/share/docutils/writers/html5_polyglot",
      ],
      "stylesheetPath": Array [
        "minimal.css",
        "plain.css",
      ],
      "tableStyle": "",
      "template": "/usr/share/docutils/writers/html5_polyglot/template.txt",
    },
    "docutilsWritersLatex2EWriter": Object {
      "attribution": "dash",
      "documentclass": "article",
      "documentoptions": "a4paper",
      "docutilsFootnotes": true,
      "embedStylesheet": 0,
      "fontEncoding": "T1",
      "footnoteReferences": "superscript",
      "graphicxOption": "",
      "hyperlinkColor": "blue",
      "hyperrefOptions": "",
      "latexPreamble": "% PDF Standard Fonts
\\\\usepackage{mathptmx} % Times
\\\\usepackage[scaled=.90]{helvet}
\\\\usepackage{courier}",
      "literalBlockEnv": "",
      "referenceLabel": null,
      "sectionEnumeratorSeparator": "-",
      "stylesheet": "",
      "stylesheetDirs": Array [
        ".",
      ],
      "tableStyle": Array [
        "standard",
      ],
      "template": "default.tex",
      "useBibtex": null,
      "useLatexAbstract": 0,
      "useLatexDocinfo": 0,
      "usePartSection": 0,
      "useVerbatimWhenPossible": 0,
    },
    "docutilsWritersOdfOdtReader": Object {
      "docinfoXform": 1,
      "doctitleXform": 1,
    },
    "docutilsWritersOdfOdtWriter": Object {
      "addSyntaxHighlighting": false,
      "cloakEmailAddresses": false,
      "createLinks": false,
      "createSections": true,
      "customFooter": "",
      "customHeader": "",
      "endnotesEndDoc": false,
      "generateOowriterToc": true,
      "stylesheet": "/usr/share/docutils/writers/odf_odt/styles.odt",
      "tableBorderThickness": null,
    },
    "docutilsWritersPepHtmlWriter": Object {
      "attribution": "dash",
      "fieldNameLimit": 14,
      "footnoteReferences": "brackets",
      "initialHeaderLevel": "1",
      "mathOutput": "HTML math.css",
      "optionLimit": 14,
      "stylesheetDirs": Array [
        ".",
        "/usr/share/docutils/writers/html4css1",
        "/usr/share/docutils/writers/html5_polyglot",
      ],
      "stylesheetPath": Array [
        "html4css1.css",
      ],
      "tableStyle": "",
      "template": "/usr/share/docutils/writers/html4css1/template.txt",
      "xmlDeclaration": 1,
    },
    "docutilsWritersS5HtmlWriter": Object {
      "attribution": "dash",
      "fieldNameLimit": 14,
      "footnoteReferences": "brackets",
      "initialHeaderLevel": "1",
      "mathOutput": "HTML math.css",
      "optionLimit": 14,
      "stylesheetDirs": Array [
        ".",
        "/usr/share/docutils/writers/html4css1",
        "/usr/share/docutils/writers/html5_polyglot",
      ],
      "stylesheetPath": Array [
        "html4css1.css",
      ],
      "tableStyle": "",
      "template": "/usr/share/docutils/writers/html4css1/template.txt",
      "xmlDeclaration": 1,
    },
    "docutilsWritersXetexWriter": Object {
      "attribution": "dash",
      "documentclass": "article",
      "documentoptions": "a4paper",
      "docutilsFootnotes": true,
      "embedStylesheet": 0,
      "footnoteReferences": "superscript",
      "graphicxOption": "",
      "hyperlinkColor": "blue",
      "hyperrefOptions": "",
      "latexPreamble": "% Linux Libertine (free, wide coverage, not only for Linux)
\\\\setmainfont{Linux Libertine O}
\\\\setsansfont{Linux Biolinum O}
\\\\setmonofont[HyphenChar=None,Scale=MatchLowercase]{DejaVu Sans Mono}",
      "literalBlockEnv": "",
      "referenceLabel": null,
      "sectionEnumeratorSeparator": "-",
      "stylesheet": "",
      "stylesheetDirs": Array [
        ".",
      ],
      "tableStyle": Array [
        "standard",
      ],
      "template": "xelatex.tex",
      "useBibtex": null,
      "useLatexAbstract": 0,
      "useLatexDocinfo": 0,
      "usePartSection": 0,
      "useVerbatimWhenPossible": 0,
    },
  },
  "source": "",
  "substitutionDefs": Object {},
  "substitutionNames": Object {},
  "symbolFootnoteRefs": Array [],
  "symbolFootnoteStart": 0,
  "symbolFootnotes": Array [],
  "tagname": "document",
  "transformMessages": Array [],
  "transformer": Transformer {
    "applied": Array [],
    "components": Object {
      "undefined": Output {
        "_init": [MockFunction],
        "encode": [MockFunction],
        "getTransforms": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "toString": [MockFunction],
        "write": [MockFunction],
      },
    },
    "document": [Circular],
    "serialno": 0,
    "sorted": 0,
    "transforms": Array [],
    "unknownReferenceResolvers": Array [],
  },
}
`;
